### PR TITLE
UCP/EP/CM: close protocol fixes

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
fix assert in close proto flush if flush is completed with transport error

### CODE REVIEW

File: src/ucp/wireup/wireup_cm.c [hunk 1]
```diff
@@ -432,12 +432,14 @@ static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
...
-           TODO: remove this case when IB flush cancel is fixed (#4743),
-                 moving QP to err state should move UCP EP to error state,
-                 then ucp_worker_set_ep_failed disconnects CM lane */
-        ucs_assert(req->status == UCS_ERR_CANCELED);
+        /* 1) ucp_ep_close(force) is called from err callback which was invoked
+              on remote connection reset
+              TODO: remove this case when IB flush cancel is fixed (#4743),
+                    moving QP to err state should move UCP EP to error state,
+                    then ucp_worker_set_ep_failed disconnects CM lane
+           2) transport level error is possible in case of > 1 lane
```
Comment (dmitrygx) [inline src/ucp/wireup/wireup_cm.c:447]: missing trailing `*/`

Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:447]: yeah, fixed...

**Review by dmitrygx — APPROVED** [review]
works for me 👍

**Comment by evgeny-leksikov** [comment]
@dmitrygx pls check last commit as well

File: src/ucp/wireup/wireup_cm.c [hunk 2]
```diff
@@ -515,19 +525,20 @@ static unsigned ucp_ep_cm_disconnect_progress(void *arg)
...
-    if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
+    if (ucp_ep->flags & UCP_EP_FLAG_FAILED) {
+        /* ignore close event on failed ep */
+        ucs_assert(!(ucp_ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID));
+    } else if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
         /* if the EP is local connected, need to flush it from main thread first */
         ucp_ep_cm_remote_disconnect_progress(ucp_ep);
         ucp_ep_invoke_err_cb(ucp_ep, UCS_ERR_CONNECTION_RESET);
-    } else if (ucp_ep->flags & UCP_EP_FLAG_CLOSED) {
+    } else {
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c]: else if (UCP_EP_FLAG_CLOSE_REQ_VALID)

**Comment by dmitrygx** [comment]
@evgeny-leksikov @yosefe tested by the following command:
```
for ((i=0;i<100000;i++)); do \
    if [ $((i%2)) -eq 0 ]; \
    then \
        off_node=sputnik4; \
    else \
        off_node=sputnik3; \
    fi; \
    echo "$i $off_node"; \
    /hpc/noarch/tmp/sputnik_switch_port_on_off.sh $off_node off; \
    /hpc/noarch/tmp/sputnik_switch_port_on_off.sh $off_node on; \
    echo "Done"; \
    sleep 30; \
done
```
after ~1 hour, it works fine with randomized IO operations and data sizes (#11)

**Pull request merged**